plugins/markdown/core/src/org/intellij/plugins/markdown/ui/preview/jcef/CodeFenceLanguageParsingSupport.kt

```diff
@@ -0,0 +1,279 @@
+// Copyright 2000-2025 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
+package org.intellij.plugins.markdown.ui.preview.jcef
+
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.startup.ProjectActivity
+import com.intellij.ui.jcef.JBCefBrowser
+import org.cef.CefSettings
+import org.cef.browser.CefBrowser
+import org.cef.handler.CefDisplayHandlerAdapter
+import org.intellij.lang.annotations.Language
+import org.intellij.markdown.ast.ASTNode
+import org.intellij.markdown.html.HtmlGenerator
+import org.intellij.plugins.markdown.ui.preview.html.DefaultCodeFenceGeneratingProvider
+import org.intellij.plugins.markdown.ui.preview.jcef.impl.executeJavaScript
+import org.intellij.plugins.markdown.ui.preview.jcef.impl.waitForPageLoad
+import java.net.URL
+import java.util.concurrent.CountDownLatch
+import java.util.concurrent.TimeUnit
+import kotlin.random.Random
+
+class CodeFenceLanguageParsingSupport : ProjectActivity {
+  override suspend fun execute(project: Project) {
+    codeFenceParsingStartUp()
+  }
+
+  companion object {
+    @Volatile private var started = false
+    @Volatile private var startupCompleted = false
+    @Volatile private var available = true
+    @Volatile private var startupLatch: CountDownLatch? = CountDownLatch(1)
+    @Volatile private var jsExecLatch: CountDownLatch? = null
+    @Volatile private var jsResultLatch: CountDownLatch? = null
+    @Volatile private var jsResult: String? = null
+    @Volatile private var jsError = false
+
+    private val browser = object : JBCefBrowser() {
```

Comment:
Please never use global state and static values. You need application service at least and better project service.

Reply:
Done. Now it's a proper application service. (I decided against a project service instead because it seemed like there would be nothing gained by initializing extra copies of the service for each opened project.)